File: cpp-client/tests/select_test.cc
/*
 * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
 */
#include <iostream>
#include <arrow/flight/client.h>
#include <arrow/flight/client_auth.h>
#include "tests/third_party/catch.hpp"
#include "tests/test_util.h"
#include "deephaven/client/client.h"
#include "deephaven/client/utility/utility.h"

#include <iostream>
#include <arrow/flight/client.h>
#include <arrow/flight/types.h>
#include <arrow/array.h>
#include <arrow/array/array_primitive.h>
#include <arrow/array/builder_binary.h>
#include <arrow/array/builder_primitive.h>
#include <arrow/compare.h>
#include <arrow/record_batch.h>
#include <arrow/scalar.h>
#include <arrow/type.h>
#include <arrow/table.h>
#include <arrow/util/key_value_metadata.h>

using deephaven::client::Client;
using deephaven::client::NumCol;
using deephaven::client::StrCol;
using deephaven::client::TableHandle;
using deephaven::client::utility::streamf;
using deephaven::client::utility::stringf;
using deephaven::client::utility::TableMaker;

namespace deephaven::client::tests {
TEST_CASE("Support all types", "[select]") {
  auto tm = TableMakerForTests::create();

  std::vector<bool> boolData;
  std::vector<char16_t> charData;
  std::vector<int8_t> byteData;
  std::vector<int16_t> shortData;
  std::vector<int32_t> intData;
  std::vector<int64_t> longData;
  std::vector<float> floatData;
  std::vector<double> doubleData;
  std::vector<std::string> stringData;

  const int startValue = -8;
  const int endValue = 8;
  for (auto i = startValue; i != endValue; ++i) {
    boolData.push_back((i % 2) == 0);
    charData.push_back(i * 10);
    byteData.push_back(i * 11);
    shortData.push_back(i * 1000);
    intData.push_back(i * 1'000'000);
    longData.push_back(i * 1'000'000'000);
    floatData.push_back(i * 123.456f);
    doubleData.push_back(i * 987654.321);
    stringData.push_back(stringf("test %o", i));
  }

  TableMaker maker;
  maker.addColumn("boolData", boolData);
  maker.addColumn("charData", charData);
  maker.addColumn("byteData", byteData);
  maker.addColumn("shortData", shortData);
  maker.addColumn("intData", intData);
  maker.addColumn("longData", longData);
  maker.addColumn("floatData", floatData);
  maker.addColumn("doubleData", doubleData);
  maker.addColumn("stringData", stringData);

  auto t = maker.makeTable(tm.client().getManager());

  std::cout << t.stream(true) << '\n';

  compareTable(
      t,
      "boolData", boolData,
      "charData", charData,
      "byteData", byteData,
      "shortData", shortData,
      "intData", intData,
      "longData", longData,
      "floatData", floatData,
      "doubleData", doubleData,
      "stringData", stringData
  );
}

TEST_CASE("Create / update / fetch a table", "[select]") {
  auto tm = TableMakerForTests::create();

  std::vector<int32_t> intData = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
  std::vector<double> doubleData = {0.0, 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9};
  std::vector<std::string> stringData = {"zero", "one", "two", "three", "four", "five", "six", "seven",
      "eight", "nine"};
  TableMaker maker;
  maker.addColumn("IntValue", intData);
  maker.addColumn("DoubleValue", doubleData);
  maker.addColumn("StringValue", stringData);
  auto t = maker.makeTable(tm.client().getManager());
  auto t2 = t.update("Q2 = IntValue * 100");
  std::cout << t2.stream(true) << '\n';
  auto t3 = t2.update("Q3 = Q2 + 10");
  std::cout << t3.stream(true) << '\n';
  auto q2 = t3.getNumCol("Q2");
  auto t4 = t3.update((q2 + 100).as("Q4"));
  std::cout << t4.stream(true) << '\n';

  std::vector<int32_t> q2Data = {0, 100, 200, 300, 400, 500, 600, 700, 800, 900};
  std::vector<int32_t> q3Data = {10, 110, 210, 310, 410, 510, 610, 710, 810, 910};
  std::vector<int32_t> q4Data = {100, 200, 300, 400, 500, 600, 700, 800, 900, 1000};

  compareTable(
      t4,
      "IntValue", intData,
      "DoubleValue", doubleData,
      "StringValue", stringData,
      "Q2", q2Data,
      "Q3", q3Data,
      "Q4", q4Data
      );
}


TEST_CASE("Select a few columns", "[select]") {
  auto tm = TableMakerForTests::create();
  auto table = tm.table();

  auto t1 = table.where("ImportDate == `2017-11-01` && Ticker == `AAPL`")
      .select("Ticker", "Close", "Volume")
      .head(2);
  std::cout << t1.stream(true) << '\n';

  // Symbolically
  auto [importDate, ticker, close, volume] =
      table.getCols<StrCol, StrCol, NumCol, NumCol>("ImportDate", "Ticker", "Close", "Volume");
  auto t2 = table.where(importDate == "2017-11-01" && ticker == "AAPL")
      .select(ticker, close, volume)
      .head(2);
  std::cout << t2.stream(true) << '\n';

  std::vector<std::string> tickerData = {"AAPL", "AAPL"};
  std::vector<double> closeData = {23.5, 24.2};
  std::vector<int64_t> volData = {100000, 250000};

  compareTable(
      t2,
      "Ticker", tickerData,
      "Close", closeData,
      "Volume", volData
      );
}

TEST_CASE("LastBy + Select", "[select]") {
  auto tm = TableMakerForTests::create();
  auto table = tm.table();

  auto t1 = table.where("ImportDate == `2017-11-01` && Ticker == `AAPL`").lastBy("Ticker")
      .select("Ticker", "Close", "Volume");
  std::cout << t1.stream(true) << '\n';

  // Symbolically
  auto importDate = table.getStrCol("ImportDate");
  auto ticker = table.getStrCol("Ticker");
  auto close = table.getNumCol("Close");
  auto volume = table.getNumCol("Volume");
  auto t2 = table.where(importDate == "2017-11-01" && ticker == "AAPL").lastBy(ticker)
      .select(ticker, close, volume);
  std::cout << t2.stream(true) << '\n';

  std::vector<std::string> tickerData = {"AAPL"};
  std::vector<double> closeData = {26.7};
  std::vector<int64_t> volData = {19000};

  compareTable(
      t2,
      "Ticker", tickerData,
      "Close", closeData,
      "Volume", volData
      );
}

TEST_CASE("New columns", "[select]") {
  auto tm = TableMakerForTests::create();
  auto table = tm.table();

  // A formula expression
  auto t1 = table.where("ImportDate == `2017-11-01` && Ticker == `AAPL`")
      .select("MV1 = Volume * Close", "V_plus_12 = Volume + 12");
  std::cout << t1.stream(true) << '\n';

  // Symbolically
  auto importDate = table.getStrCol("ImportDate");
  auto ticker = table.getStrCol("Ticker");
  auto close = table.getNumCol("Close");
  auto volume = table.getNumCol("Volume");

  auto t2 = table.where(importDate == "2017-11-01" && ticker == "AAPL")
      .select((volume * close).as("MV1"), (volume + 12).as("V_plus_12"));
  std::cout << t2.stream(true) << '\n';

  std::vector<double> mv1Data = {2350000, 6050000, 507300};
  std::vector<int64_t> mv2Data = {100012, 250012, 19012};

  const TableHandle *tables[] = {&t1, &t2};
  for (const auto *t : tables) {
    compareTable(
        *t,
        "MV1", mv1Data,
        "V_plus_12", mv2Data
    );
  }
}

TEST_CASE("Simple where", "[select]") {
  auto tm = TableMakerForTests::create();
  auto table = tm.table();
  auto updated = table.update("QQQ = i");
  // Symbolically
  auto importDate = updated.getStrCol("ImportDate");
  auto ticker = updated.getStrCol("Ticker");
  auto volume = updated.getNumCol("Volume");
  // if we allowed C++17 we could do something like
  // auto [importDate, ticker, volume] = table.getCol<StrCol, StrCol, NumCol>("ImportDate", "Ticker", "Volume");

  auto t2 = updated.where(importDate == "2017-11-01" && ticker == "IBM")
      .select(ticker, volume);
  std::cout << t2.stream(true) << '\n';

  std::vector<std::string> tickerData = {"IBM"};
  std::vector<int64_t> volData = {138000};

  compareTable(
      t2,
      "Ticker", tickerData,
      "Volume", volData
  );
}

TEST_CASE("Formula in the where clause", "[select]") {
  auto tm = TableMakerForTests::create();
  auto table = tm.table();

  auto t1 = table.where("ImportDate == `2017-11-01` && Ticker == `AAPL` && Volume % 10 == Volume % 100")
      .select("Ticker", "Volume");
  std::cout << t1.stream(true) << '\n';

  // Symbolically
  auto importDate = table.getStrCol("ImportDate");
  auto ticker = table.getStrCol("Ticker");
  auto volume = table.getNumCol("Volume");
  auto t2 = table.where(importDate == "2017-11-01" && ticker == "AAPL" && volume % 10 == volume % 100)
      .select(ticker, volume);
  std::cout << t2.stream(true) << '\n';

  std::vector<std::string> tickerData = {"AAPL", "AAPL", "AAPL"};
  std::vector<int64_t> volData = {100000, 250000, 19000};

  const TableHandle *tables[] = {&t1, &t2};
  for (const auto *t : tables) {
    compareTable(
        *t,
        "Ticker", tickerData,
        "Volume", volData
        );
  }
}

TEST_CASE("Simple 'where' with syntax error", "[select]") {
  auto tm = TableMakerForTests::create();
  auto table = tm.table();

  try {
    // String literal
    auto t1 = table.where(")))))");
    std::cout << t1.stream(true) << '\n';
  } catch (const std::exception &e) {
    // Expected
    streamf(std::cerr, "Caught *expected* exception %o\n", e.what());
    return;
  }
  throw std::runtime_error("Expected a failure, but didn't experience one");
}
}  // namespace deephaven::client::tests

File: csharp/client/DhClientTests/SelectTest.cs
using Deephaven.DeephavenClient.Interop;
using System;
using Deephaven.DeephavenClient;
using Deephaven.DeephavenClient.Utility;
using Xunit.Abstractions;

namespace Deephaven.DhClientTests;

public class SelectTest {
  private readonly ITestOutputHelper _output;

  public SelectTest(ITestOutputHelper output) {
    _output = output;
  }

  [Fact]
  public void TestSupportAllTypes() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());

    var boolData = new List<bool>();
    var charData = new List<char>();
    var byteData = new List<sbyte>();
    var shortData = new List<short>();
    var intData = new List<int>();
    var longData = new List<long>();
    var floatData = new List<float>();
    var doubleData = new List<double>();
    var stringData = new List<string>();
    var dateTimeData = new List<DhDateTime>();

    const int startValue = -8;
    const int endValue = 8;
    for (var i = startValue; i != endValue; ++i) {
      boolData.Add((i % 2) == 0);
      charData.Add((char)(i * 10));
      byteData.Add((sbyte)(i * 11));
      shortData.Add((short)(i * 1000));
      intData.Add(i * 1_000_000);
      longData.Add(i * (long)1_000_000_000);
      floatData.Add(i * 123.456F);
      doubleData.Add(i * 987654.321);
      stringData.Add($"test {i}");
      dateTimeData.Add(DhDateTime.FromNanos(i));
    }

    using var maker = new TableMaker();
    maker.AddColumn("boolData", boolData);
    maker.AddColumn("charData", charData);
    maker.AddColumn("byteData", byteData);
    maker.AddColumn("shortData", shortData);
    maker.AddColumn("intData", intData);
    maker.AddColumn("longData", longData);
    maker.AddColumn("floatData", floatData);
    maker.AddColumn("doubleData", doubleData);
    maker.AddColumn("stringData", stringData);
    maker.AddColumn("dateTimeData", dateTimeData);

    var t = maker.MakeTable(ctx.Client.Manager);

    _output.WriteLine(t.ToString(true));

    var tc = new TableComparer();
    tc.AddColumn("boolData", boolData);
    tc.AddColumn("charData", charData);
    tc.AddColumn("byteData", byteData);
    tc.AddColumn("shortData", shortData);
    tc.AddColumn("intData", intData);
    tc.AddColumn("longData", longData);
    tc.AddColumn("floatData", floatData);
    tc.AddColumn("doubleData", doubleData);
    tc.AddColumn("stringData", stringData);
    tc.AddColumn("dateTimeData", dateTimeData);
    tc.AssertEqualTo(t);
  }

  [Fact]
  public void TestCreateUpdateFetchATable() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());

    var intData = new[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
    var doubleData = new[] { 0.0, 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9 };
    var stringData = new[] {
      "zero", "one", "two", "three", "four", "five", "six", "seven",
      "eight", "nine"
    };

    using var maker = new TableMaker();
    maker.AddColumn("IntValue", intData);
    maker.AddColumn("DoubleValue", doubleData);
    maker.AddColumn("StringValue", stringData);
    var t = maker.MakeTable(ctx.Client.Manager);
    var t2 = t.Update("Q2 = IntValue * 100");
    var t3 = t2.Update("Q3 = Q2 + 10");
    var t4 = t3.Update("Q4 = Q2 + 100");

    var q2Data = new[] { 0, 100, 200, 300, 400, 500, 600, 700, 800, 900 };
    var q3Data = new[] { 10, 110, 210, 310, 410, 510, 610, 710, 810, 910 };
    var q4Data = new[] { 100, 200, 300, 400, 500, 600, 700, 800, 900, 1000 };

    var tc = new TableComparer();
    tc.AddColumn("IntValue", intData);
    tc.AddColumn("DoubleValue", doubleData);
    tc.AddColumn("StringValue", stringData);
    tc.AddColumn("Q2", q2Data);
    tc.AddColumn("Q3", q3Data);
    tc.AddColumn("Q4", q4Data);
    tc.AssertEqualTo(t4);
  }

  [Fact]
  public void TestSelectAFewColumns() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;

    var t1 = table.Where("ImportDate == `2017-11-01` && Ticker == `AAPL`")
      .Select("Ticker", "Close", "Volume")
      .Head(2);

    var tickerData = new[] { "AAPL", "AAPL" };
    var closeData = new[] { 23.5, 24.2 };
    var volData = new Int64[] { 100000, 250000 };

    var tc = new TableComparer();
    tc.AddColumn("Ticker", tickerData);
    tc.AddColumn("Close", closeData);
    tc.AddColumn("Volume", volData);
    tc.AssertEqualTo(t1);
  }

  [Fact]
  public void TestLastByAndSelect() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;

    var t1 = table.Where("ImportDate == `2017-11-01` && Ticker == `AAPL`")
      .LastBy("Ticker")
      .Select("Ticker", "Close", "Volume");
    _output.WriteLine(t1.ToString(true));

    var tickerData = new[] { "AAPL" };
    var closeData = new[] { 26.7 };
    var volData = new Int64[] { 19000 };

    var tc = new TableComparer();
    tc.AddColumn("Ticker", tickerData);
    tc.AddColumn("Close", closeData);
    tc.AddColumn("Volume", volData);
    tc.AssertEqualTo(t1);
  }

  [Fact]
  public void TestNewColumns() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;

    // A formula expression
    var t1 = table.Where("ImportDate == `2017-11-01` && Ticker == `AAPL`")
      .Select("MV1 = Volume * Close", "V_plus_12 = Volume + 12");

    var mv1Data = new double[] { 2350000, 6050000, 507300 };
    var vp2Data = new long[] { 100012, 250012, 19012 };

    var tc = new TableComparer();
    tc.AddColumn("MV1", mv1Data);
    tc.AddColumn("V_plus_12", vp2Data);
    tc.AssertEqualTo(t1);
  }

  [Fact]
  public void TestDropColumns() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;

    var t1 = table.DropColumns("ImportDate", "Open", "Close");
    Assert.Equal(5, table.Schema.NumCols);
    Assert.Equal(2, t1.Schema.NumCols);
  }

  [Fact]
  public void TestSimpleWhere() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;
    var updated = table.Update("QQQ = i");

    var t1 = updated.Where("ImportDate == `2017-11-01` && Ticker == `IBM`")
      .Select("Ticker", "Volume");

    var tickerData = new[] { "IBM" };
    var volData = new Int64[] { 138000 };

    var tc = new TableComparer();
    tc.AddColumn("Ticker", tickerData);
    tc.AddColumn("Volume", volData);
    tc.AssertEqualTo(t1);
  }

  [Fact]
  public void TestFormulaInTheWhereClause() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;

    var t1 = table.Where(
        "ImportDate == `2017-11-01` && Ticker == `AAPL` && Volume % 10 == Volume % 100")
      .Select("Ticker", "Volume");
    _output.WriteLine(t1.ToString(true));

    var tickerData = new[] { "AAPL", "AAPL", "AAPL" };
    var volData = new Int64[] { 100000, 250000, 19000 };

    var tc = new TableComparer();
    tc.AddColumn("Ticker", tickerData);
    tc.AddColumn("Volume", volData);
    tc.AssertEqualTo(t1);
  }

  [Fact]
  public void TestSimpleWhereWithSyntaxError() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var table = ctx.TestTable;

    Assert.Throws<Exception>(() => {
      var t1 = table.Where(")))))");
      _output.WriteLine(t1.ToString(true));
    });
  }

  [Fact]
  public void TestWhereIn() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());

    var letterData = new[] { "A", "C", "F", "B", "E", "D", "A" };
    var numberData = new Int32?[] { null, 2, 1, null, 4, 5, 3 };
    var colorData = new[] { "red", "blue", "orange", "purple", "yellow", "pink", "blue" };
    var codeData = new Int32?[] { 12, 13, 11, null, 16, 14, null };

    using var sourceMaker = new TableMaker();
    sourceMaker.AddColumn("Letter", letterData);
    sourceMaker.AddColumn("Number", numberData);
    sourceMaker.AddColumn("Color", colorData);
    sourceMaker.AddColumn("Code", codeData);

    var source = sourceMaker.MakeTable(ctx.Client.Manager);

    var filterColorData = new[] { "blue", "red", "purple", "white" };

    using var filterMaker = new TableMaker();
    filterMaker.AddColumn("Colors", filterColorData);
    var filter = filterMaker.MakeTable(ctx.Client.Manager);

    var result = source.WhereIn(filter, "Color = Colors");

    var letterExpected = new[] { "A", "C", "B", "A" };
    var numberExpected = new Int32?[] { null, 2, null, 3 };
    var colorExpected = new[] { "red", "blue", "purple", "blue" };
    var codeExpected = new Int32?[] { 12, 13, null, null };

    var expected = new TableComparer();
    expected.AddColumn("Letter", letterExpected);
    expected.AddColumn("Number", numberExpected);
    expected.AddColumn("Color", colorExpected);
    expected.AddColumn("Code", codeExpected);

    expected.AssertEqualTo(result);
  }

  [Fact]
  public void TestLazyUpdate() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());

    var aData = new[] { "The", "At", "Is", "On" };
    var bData = new[] { 1, 2, 3, 4 };
    var cData = new[] { 5, 2, 5, 5 };

    using var sourceMaker = new TableMaker();
    sourceMaker.AddColumn("A", aData);
    sourceMaker.AddColumn("B", bData);
    sourceMaker.AddColumn("C", cData);
    var source = sourceMaker.MakeTable(ctx.Client.Manager);

    var result = source.LazyUpdate("Y = sqrt(C)");

    var sqrtData = new[] { Math.Sqrt(5), Math.Sqrt(2), Math.Sqrt(5), Math.Sqrt(5) };

    var tc = new TableComparer();
    tc.AddColumn("A", aData);
    tc.AddColumn("B", bData);
    tc.AddColumn("C", cData);
    tc.AddColumn("Y", sqrtData);
    tc.AssertEqualTo(result);
  }

  [Fact]
  public void TestSelectDistinct() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());

    var aData = new[] { "apple", "apple", "orange", "orange", "plum", "grape" };
    var bData = new[] { 1, 1, 2, 2, 3, 3 };

    using var sourceMaker = new TableMaker();
    sourceMaker.AddColumn("A", aData);
    sourceMaker.AddColumn("B", bData);
    var source = sourceMaker.MakeTable(ctx.Client.Manager);

    var result = source.SelectDistinct("A");
    _output.WriteLine(result.ToString(true));

    var expectedAData = new[] { "apple", "orange", "plum", "grape" };

    var tc = new TableComparer();
    tc.AddColumn("A", expectedAData);
    tc.AssertEqualTo(result);
  }
}
